### PR TITLE
Generic ppx driver

### PR DESCRIPTION
**Pull request by diml**
This PR removes the hard-coded knowledge of ppx_driver and ocaml-migrate-parsetree and their specific details. Instead we know have this in the jbuild files of ocaml-migrate-parsetree and ppx_driver:

```scheme
(library
 ((name migrate_parsetree)
  [...]
  (ppx.driver
   ((main       Migrate_parsetree.Driver.run_main)
    (flags      (--dump-ast))
    (lint_flags (--null))))))

(library
 ((name ppx_driver)
  [...]
  (ppx.driver
   ((main       Ppx_driver.standalone)
    (replaces   (ocaml-migrate-parsetree))
    (flags      (-corrected-suffix ${corrected-suffix} -diff-cmd - -dump-ast))
    (lint_flags (-corrected-suffix ${corrected-suffix} -diff-cmd - -null    ))))
  ))
```

This makes things a bit cleaner and in particular we no longer have to synchronise the releases of Dune and these packages when adding new features.

To test, it is necessary to use the `generic-ppx-driver` branch of ocaml-migrate-parsetree and ppx_driver

### CODE REVIEW

**Comment by rgrinberg** [comment]
Looks good but it's a shame the error when the driver is missing is so opaque. Previously, we'd tell the user they needed to install something. It would be nice to have something more helpful than "No ppx driver found"

**Comment by diml** [comment]
BTW, I'm trying to think what is the best way to release this. We need to add dependency constraints on released ppx_driver and ocaml-migrate-parsetree packages, but then we have the choice between:

option1:

- releasing jbuilder beta19
- releasing new versions of ppx_driver and ocaml-migrate-parsetree

and in between, users won't be able to upgrade to jbuilder beta19

option 2:

- releasing jbuilder beta19 that only accept the new `ppx.driver` field but ignores it
- releasing a version of ppx_driver and ocaml-migrate-parsetree with the updated jbuild files
- releasing jbuilder beta20 with this PR in

**Comment by rgrinberg** [comment]
I think option #1 is doable if we batch together the releases of jbuilder, omp, and driver. But we'd also like to avoid delaying beta19 as it contains some important fixes. So unless those other releases are imminent, I'd favour #2.

**Comment by diml** [comment]
Yh, the other releases are ready

**Comment by diml** [comment]
But maybe 2 is better anyway. For instance the latest ppx_driver is not compatible with older version of OCaml so we will need multiple minor releases of ppx_driver.

**Comment by rgrinberg** [comment]
Let's go with #2 then

File: src/inline_tests.ml [hunk 1]
```diff
@@ -56,32 +57,31 @@ module Backend = struct
...
-                     | Some t -> Ok t)))
+          Result.all
+            (List.map info.extends
+               ~f:(fun ((loc, name) as x) ->
+                 resolve x >>= fun lib ->
+                 match get ~loc lib with
+                 | None ->
+                   Error (Loc.exnf loc "%S is not an %s" name
+                            (desc ~plural:false))
+                 | Some t -> Ok t))
```
Comment (rgrinberg) [inline src/inline_tests.ml]: Would it be possible to move these changes into their own PR if possible?

Comment (diml) [reply src/inline_tests.ml]: Ok, I'll cherry-pick directly some of the simpler commit to master directly, since we can't create dependent PRs on github

Comment (diml) [reply src/inline_tests.ml]: Done: https://github.com/ocaml/dune/pull/632

**Comment by rgrinberg** [comment]
@diml what's the status of this? You think you'll have enough time to polish this for the next release?

**Comment by diml** [comment]
I improved the error messages, now it suggests to upgrade ocaml-migrate-parsetree, ppxlib or ppx_driver if the pps depend on them. Otherwise it suggests that they are not compatible with jbuilder.

**Comment by diml** [comment]
But we need to wait for a new version of ppxlib

**Comment by diml** [comment]
The recent problems we had with ppx_sexp_conv makes me think that this is going to cause more pain than I expected. I think I will change this feature so that the new behavior is only activated in dune projects. This way nothing will change for existing released projects.

**Comment by diml** [comment]
I added a commit on top of the previous changes to restore the old behavior when the file is a jbuild file.  To know if the file is a jbuild file, we pass a `dir_kind` value around that is either `Dune` or `Jbuild`. The *old* ppx drivers are stored in `_build/default/.ppx/jbuild`.

The code is not exactly the same as what we had before, but it should work for all existing projects.

We shouldn't need to do anything special for the next release. Existing projects will continue to work, and if there is an issue with a pure dune project because ppxlib or omp was reinstalled, Dune prints an error message inviting the user to reinstall omp or ppxlib.

@rgrinberg could you review the last changes? I think this is now safe to merge.

File: src/install_rules.ml [hunk 2]
```diff
@@ -184,25 +184,30 @@ module Gen(P : Install_params) = struct
...
-                Loc.none, Pp.of_string "ppx_driver.runner"]
-          else
-            pps
-        in
-        let ppx_exe = Preprocessing.get_ppx_driver sctx ~scope pps in
-        [ppx_exe]
+          match Preprocessing.get_ppx_driver sctx ~scope ~dir_kind pps with
+          | Ok    x -> [x]
+          | Error _ ->
+            [Preprocessing.get_ppx_driver_for_public_lib sctx ~name ~dir_kind]
```
Comment (rgrinberg) [inline src/install_rules.ml]: Could you explain this bit? What kind of errors are we ignoring here?

**Comment by rgrinberg** [comment]
Those last (recent) commits look good. Btw, how would this work for mixed setups with both dune and jbuild files? So we're not going to be able to use both the new and old drivers in the same workspace?

File: src/install_rules.ml (hunk 2 above)
Comment (diml) [reply src/install_rules.ml]: Looking at the code again, I believe the only possible failure is if `ppxlib.runner` or `ppx_driver.runner` doesn't exist. This might be the case at some point in the future and this code is not very clear anyway. I reworked it to make things more explicit and avoid the error case here.

**Comment by diml** [comment]
It should work fine. Basically when we are in a jbuild directory, the new driver stuff is not used at all and we still use the hardcoded values. The old drivers are built in a separate directory `.ppx/jbuild`, so it should be possible to use both method without issues.

File: src/preprocessing.ml [hunk 3]
```diff
@@ -203,7 +203,9 @@ let no_driver_error pps =
   | None ->
     sprintf
       "No ppx driver found.\n\
-       It seems that these ppx rewriters are not compatible with jbuilder."
+       It seems that these ppx rewriters are not compatible with Dune.\n\
```
Comment (rgrinberg) [inline src/preprocessing.ml]: Rather than "these", should we just include the pps here in the error message (which also has locs)

Comment (diml) [reply src/preprocessing.ml]: Indeed. We don't always have the location at this point because we might reach this program point when generating the rules for `_build/default/.ppx/<key>/ppx.exe`.

I'm refactoring a bit this code because it's becoming messy.

**Comment by diml** [comment]
I improved the error messages and added tests.

**Comment by diml** [comment]
I'm glad to have finished this one!
